### DISCUSSION TITLE
Inconsistency using foreign object with D3 in Chromium

### DISCUSSION THREAD

**Discussion by helderdarocha** [d3/d3#3956] [Q&A] [answered]
I converted an existing DOM code to the D3 code below and it stopped working as before, but only in Chromium. Clicking the link should open the page in the `iframe`. Safari and Chrome can't find the `target` and open in the link a new window/tab. It works as expected in Firefox:
```
<script type="module">
    import * as d3 from "https://cdn.skypack.dev/d3@7";

    const svg = d3.select("body")
                  .append("svg")
                  .attr("width", 825)
                  .attr("height", 375);

    svg.append("a")
       .attr("href", "https://d3js.org/")
       .attr("target", "result")
       .append("text")
           .attr("x", 50)
           .attr("y", 175)
           .text("Click me");

    svg.append("foreignObject")
       .attr("x", 150)
       .attr("y", 50)
       .attr("width", 606)
       .attr("height", 256)
       .append("xhtml:iframe")
            .attr("width", 600)
            .attr("height", 250)
            .attr("name", "result");
</script>
```
This generates the following code, which works everywhere if generated via DOM:
```
<a href="https://d3js.org/" target="result"><text x="50" y="175">Click me</text></a>
<foreignObject x="150" y="50" width="105" height="55">
        <iframe width="100" height="50" name="result"><html><head></head><body></body></html></iframe>
</foreignObject>
```
I'm not sure if this is a D3 bug or a Chromium bug, but it seems to be related to the way namespaces are handled.

**Comment by Fil** [comment]
I can't reproduce the issue. It works for me in Chrome, Safari, Chromium (Version 123.0.6312.105 (Official Build, ungoogled-chromium) (arm64)).

**Reply by helderdarocha** [reply]
I tried it in Chrome Version 131.0.6778.109 (Official Build) (arm64), Firefox 133.0 (aarch64), and Safari Version 17.6 (19618.3.11.11.5), all running on a MacBook Air M2 2022 with Sonoma 14.6.1. In Firefox the links open in the iframe. In Safari and Chrome they open a new tab. This only happens when the code is generated with D3 with the code above. It works in static SVG and DOM-generated SVG.

**Reply by Fil** [reply]
I have macos 15.1.1 but I can't imagine it makes much of difference. Do you have the issue on [this page](https://observablehq.com/@fil/discussion-3956)? Can you share a page where you see the issue?

**Reply by helderdarocha** [reply]
In the link you posted it works. I copied the code [here](https://codepen.io/helderdarocha/pen/yyBeEQB) (CodePen) and the issue happens in Chrome, but not in Firefox.

**Reply by Fil** [reply]
codepen is not really "a page" — it creates an iframe with your page, and does a lot of wrapping/renaming around it for security.

**Reply by helderdarocha** [reply]
Yes. I know. But since I was able to reproduce this issue there, I thought the link would be useful.

**Reply by helderdarocha** [reply]
I also placed it [here](https://stelle.com.br/tmp/foreignobject-d3-bug.html), which is a normal web page :)

**Reply by Fil** [reply]
Thanks, this is helpful. I see that we have the problem if we append the svg element to the body then append the link and the iframe; and no bug if we create everything and append the svg element afterwards. This also explains why it worked in my notebook.

I've now added several cells which insert the elements at different moments, and that seems to show that the issue happens when the target iframe is not yet in the dom when we add the link. I think it demonstrates a browser bug.

**Reply by helderdarocha** [reply]
That's probably why it works in DOM version [here](https://www.stelle.com.br/tmp/foreignobject-d3-bug-DOM.html), since all the elements are first created and then appended.

I found another of these bugs (works in DOM+any browser, fails in Chrome+D3) related to animations. I solved it for my client with D3 transitions, but the bug exists. I will try to create a small version of it and post a new comment about it.

**Comment by helderdarocha** [comment, marked as answer]
This is a workaround, as suggested by Philippe Rivière:
```
<script type="module">
    import * as d3 from "https://cdn.skypack.dev/d3@7";

    const detachedSVG = d3.create("svg")        // create a detached SVG
                          .attr("width", 825)
                          .attr("height", 375);

    detachedSVG.append("a")
       .attr("href", "https://d3js.org/")
       .attr("target", "result")
       .append("text")
           .attr("x", 50)
           .attr("y", 175)
           .text("Click me");

    detachedSVG.append("foreignObject")
       .attr("x", 150)
       .attr("y", 50)
       .attr("width", 606)
       .attr("height", 256)
       .append("xhtml:iframe")
            .attr("width", 600)
            .attr("height", 250)
            .attr("name", "result");

    d3.select("body").append(() => detachedSVG.node());    // append to root after adding foreign object
</script>
```

**Reply by helderdarocha** [reply]
Another example. This time the `begin="click"` attribute in `<animate>` is ignored, and, in Chromium browsers, the animation starts as soon as the page loads, not after the user clicks the rectangle:
```
const svg = d3.select("body")
              .append("svg")
              .attr("width", 800)
              .attr("height", 400);

svg.append("rect").attr("x", 50).attr("y", 50)
      .attr("width", 100).attr("height", 50)
      .attr("fill", "yellow")
      .append("animate")
            .attr("attributeName", "fill")
            .attr("to", "purple")
            .attr("dur", "3s")
            .attr("begin", "click")
            .attr("fill", "freeze");
```
The solution is also to create the SVG first and append it after:
```
const svg = d3.create("svg")
              .attr("width", 800)
              .attr("height", 200);

/* add descendants to svg */

d3.select("body").append(() => svg.node());
```
